pennylane/ops/PauliArithmetic.py

```diff
@@ -0,0 +1,194 @@
+}
+
+
+class PauliWord(dict):
+    """Immutable dictionary used to represent a Pauli Word.
+    Can be constructed from a standard dictionary.
+
+    >>> w = PauliWord({"a": X, 2: Y, 3: z})
+    """
+    _map_I = {
+        I: (1, I),
+        X: (1, X),
+        Y: (1, Y),
+        Z: (1, Z),
+    }
+    _map_X = {
+        I: (1, X),
+        X: (1, I),
+        Y: (1.0j, Z),
+        Z: (-1.0j, Y),
+    }
+    _map_Y = {
+        I: (1, Y),
+        X: (-1.0j, Z),
+        Y: (1, I),
+        Z: (1j, X),
+    }
+    _map_Z = {
+        I: (1, Z),
+        X: (1j, Y),
+        Y: (-1.0j, X),
+        Z: (1, I),
+    }
+
+    mul_map = {
+        I: _map_I,
+        X: _map_X,
+        Y: _map_Y,
+        Z: _map_Z
+    }
+
+    def __missing__(self, key):
+        """If the wire is not in the Pauli word,
+        then no operator acts on it, so return the Identity."""
+        return I
+
+    def __setitem__(self, key, item):
+        raise NotImplementedError
+
+    def __hash__(self):
+        return hash(frozenset(self.items()))
+
+    def __mul__(self, other):
+        d = dict(self)
+        coeff = 1
+
+        for wire, term in other.items():
+
+            if wire in d:
+                factor, new_op = self.mul_map[d[wire]][term]
+                if new_op == I:
+                    del d[wire]
+                else:
+                    coeff *= factor
+                    d[wire] = new_op
+            elif term != I:
+                d[wire] = term
+
+        return PauliWord(d), coeff
+
+
+class PauliSentence(dict):
+    """Dict representing a Pauli Sentence."""
+    def __missing__(self, key):
+        """If the pauliword is not in the sentence then the coefficient
+        associated with it should be 0."""
+        return 0.0
+
+    def __add__(self, other):
+        """Add two Pauli sentence together by iterating over the smaller
+        one and adding its terms to the larger one."""
+        smaller_ps, larger_ps = (self, other) if len(self) < len(other) else (other, self)
+        for key in smaller_ps:
+            larger_ps[key] += smaller_ps[key]
+
+        return larger_ps
+
+    def __mul__(self, other):
+        """Multiply two Pauli sentences by iterating over each sentence and multiplying
+        the Pauli words pair-wise"""
+        final_ps = PauliSentence({})
+        for pw1 in self:
+            for pw2 in other:
+                prod_pw, coeff = pw1 * pw2
+                final_ps[prod_pw] += coeff * self[pw1] * other[pw2]
+
+        return final_ps
+
+    def __str__(self):
+        rep_str = ""
+        for index, (pw, coeff) in enumerate(self.items()):
+            if index == 0:
+                rep_str += "= "
+            else:
+                rep_str += "+ "
+            rep_str += f"({round(coeff, 2)}) * "
+            for w, op in pw.items():
+                rep_str += f"[{op}({w.labels[0]})]"
+            rep_str += "\n"
+
+        return rep_str
+
+    def _to_mat(self, wire_order, sparse=False):
```

Comment:
Maybe instead of `sparse=False`/ `sparse=True`, we could have `format="dense"`, `format="csr"`, `format='coo'`, etc.

Reply:
Added 👍🏼